File: JavaScript/815-Bus Routes.js
/**
 * @param {number[][]} routes
 * @param {number} source
 * @param {number} target
 * @return {number}
 */
var numBusesToDestination = function(routes, source, target) {
    let stopToRoute = new Map();
    for (let [routeIdx, route] of routes.entries()){
        for (let stop of route){
            stopToRoute.set(stop, [routeIdx].concat(stopToRoute.get(stop) || []) );
        }
    }
    
    let visitedBus = new Set();
    let queue = [[source, 0]];
    
    while (queue.length){
        let [currStop, count] = queue.shift();
        if (currStop === target){
            return count;
        }
        for(let routeIdx of stopToRoute.get(currStop)){
            if (!visitedBus.has(routeIdx)){
                visitedBus.add(routeIdx);
                for (let nxtStop of routes[routeIdx]){
                    queue.push([nxtStop, count+1]);
                }
            }
        }
    }
    return -1;
};

File: C++/815-Bus Routes.cpp
class Solution {
public:
    int numBusesToDestination(vector<vector<int>>& routes, int source, int target) {
        
        unordered_map<int, vector<int>> group;
        for (int i=0; i<routes.size(); i++){
            for (int x:routes[i]){
                group[x].push_back(i);
            }
        }
        
        queue<int> q;
        q.push(source);
        unordered_set<int> visitedGroup;
        unordered_set<int> visitedStop;
        visitedStop.insert(source);
        
        int step = 0;
        
        while (!q.empty()){
            int size = q.size();
            for (int i=0; i<size; i++){
                int curr = q.front(); q.pop();
                if (curr == target) return step;
                for (int groupIdx:group[curr]){
                    if (!visitedGroup.count(groupIdx)){
                        visitedGroup.insert(groupIdx);
                        for (int stop:routes[groupIdx]){
                            if (!visitedStop.count(stop)){
                                visitedStop.insert(stop);
                                q.push(stop);
                            }
                        }
                    }
                }
            }
            step++;
        }
        return -1;
    }
};